### PR TITLE
Implement Features Needed for DbDataAdapter

### LINKED ISSUE #16: Implement Features Needed for DbDataAdapter
## Sample Code
```c#
	var mockConnection = Substitute.For<DbConnection>().SetupCommands();
	// ToDo: SetupQuery
	var mockAdapter = Substitute.ForPartsOf<DbDataAdapter>();
	mockAdapter.SelectCommand = mockConnection.CreateCommand();
	mockAdapter.SelectCommand.CommandText = "SELECT * FROM table1";
	mockAdapter.Fill(new DataSet());
```

## Expected Outcome
Fill executes the mocked query

## Actual Result
Fill: SelectCommand.Connection property has not been initialized.
   at System.Data.Common.DbDataAdapter.GetConnection3(DbDataAdapter adapter, IDbCommand command, String method)
   at System.Data.Common.DbDataAdapter.FillInternal(DataSet dataset, DataTable[] datatables, Int32 startRecord, Int32 maxRecords, String srcTable, IDbCommand command, CommandBehavior behavior)
   at System.Data.Common.DbDataAdapter.Fill(DataSet dataSet, Int32 startRecord, Int32 maxRecords, String srcTable, IDbCommand command, CommandBehavior behavior)
   at Castle.Proxies.DbDataAdapterProxy.Fill_callback_1(DataSet dataSet, Int32 startRecord, Int32 maxRecords, String srcTable, IDbCommand command, CommandBehavior behavior)
   at Castle.Proxies.Invocations.DbDataAdapter_Fill_1.InvokeMethodOnTarget()
   at Castle.DynamicProxy.AbstractInvocation.Proceed()
   at NSubstitute.Proxies.CastleDynamicProxy.CastleInvocationMapper.<>c__DisplayClass4_0.<CreateBaseResultInvocation>b__0()

### PR DESCRIPTION
**Pull request by microalps**
Resolves #16

### CODE REVIEW

**Comment by edovbenko** [comment]
looking forward to see this PR merged

File: NSubstitute.DbConnection/NSubstitute.DbConnection.Tests/QueryTests.cs [hunk 1]
```diff
@@ -7,21 +7,30 @@
...
 namespace NSubstitute.DbConnection.Tests;
 
+using System.Data;
 using System.Data.Common;
 
 public class QueryTests
 {
-    [Test]
-    public void ShouldMockQuery()
+    private const string SimpleSelect = "select * from table";
```
Comment (jmg48) [inline NSubstitute.DbConnection/NSubstitute.DbConnection.Tests/QueryTests.cs]: This is only used inside the private method so I would make it local within the method

File: NSubstitute.DbConnection/NSubstitute.DbConnection.Tests/QueryTests.cs [hunk 2]
```diff
@@ -7,21 +7,30 @@
...
+
+    private static DbCommand GetSimpleCommandSingleRow(Action<DbConnection> open)
```
Comment (jmg48) [inline NSubstitute.DbConnection/NSubstitute.DbConnection.Tests/QueryTests.cs]: Can we move this to the end of the class - I know I don't have method order linting enabled but prefer public before private

File: NSubstitute.DbConnection/NSubstitute.DbConnection/MockQuery.cs [hunk 3]
```diff
@@ -148,6 +148,16 @@ public DbDataReader ExecuteReader(IDbCommand mockCommand)
             mockReader[Arg.Any<int>()].Returns(ci => GetValue(FieldByIndex((int)ci[0])));
             mockReader.GetValue(Arg.Any<int>()).Returns(ci => GetValue(FieldByIndex((int)ci[0])));
             mockReader[Arg.Any<string>()].Returns(ci => GetValue(FieldByName((string)ci[0])));
+            mockReader.GetValues(Arg.Any<object[]>()).Returns(ci =>
+            {
+                var values = ci.ArgAt<object[]>(0);
+                for (var fieldIndex = 0; fieldIndex < values.Length; fieldIndex++)
```
Comment (jmg48) [inline NSubstitute.DbConnection/NSubstitute.DbConnection/MockQuery.cs]: Looks like this would fail out of range for certain wrongly-sized arrays. I'd be happy to assert the array needs to be the correct size here and just fail with a more descriptive error (i.e. be opinionated here rather than try to mock the edge cases), or otherwise we could handle and have test cases for

1. array the same width as the result set
2. array narrower than the result set
3. array wider that the result set

**Review by jmg48 — CHANGES_REQUESTED** [review]
I'm happy to merge this as is and then tidy up around those comments as I prep the next package release, or equally happy for you to address the comments - just let me know!

File: NSubstitute.DbConnection/NSubstitute.DbConnection/DbConnectionWrapper.cs [hunk 4]
```diff
@@ -54,7 +54,12 @@ public IMockQueryBuilder AddQuery(Func<string, bool> matcher)
 
         protected override DbTransaction BeginDbTransaction(IsolationLevel isolationLevel) => (DbTransaction)_inner.BeginTransaction();
 
-        protected override DbCommand CreateDbCommand() => (DbCommand)_inner.CreateCommand();
+        protected override DbCommand CreateDbCommand()
+        {
+            var cmd = (DbCommand)_inner.CreateCommand();
+            cmd.Connection = this;
```
Comment (jmg48) [inline NSubstitute.DbConnection/NSubstitute.DbConnection/DbConnectionWrapper.cs]: Suggest instead that this is added above [here](https://github.com/jmg48/NSubstitute.DbConnection/blob/main/NSubstitute.DbConnection/NSubstitute.DbConnection/DbConnnectionMockExtensions.cs#L124) as `mockCommand.Connection.Returns(result);`

File: NSubstitute.DbConnection/NSubstitute.DbConnection.Tests/QueryTests.cs (hunk 1 above)
Comment (microalps) [reply NSubstitute.DbConnection/NSubstitute.DbConnection.Tests/QueryTests.cs]: Was intended to replace other methods that use the same select, but I'm fine with this suggestion and this can always be done later.

File: NSubstitute.DbConnection/NSubstitute.DbConnection/MockQuery.cs (hunk 3 above)
Comment (microalps) [reply NSubstitute.DbConnection/NSubstitute.DbConnection/MockQuery.cs]: I used ODBC [Reference Source](https://referencesource.microsoft.com/#System.Data/fx/src/data/System/Data/Odbc/OdbcDataReader.cs,502) as the model to properly handle this scenario.

**Review by jmg48 — APPROVED** [review]
Looks great, many thanks!

**Comment by jmg48** [comment]
Pushed to nuget as 2.14, should be available shortly
